### PR TITLE
event loop notifications to websocket clients

### PR DESCRIPTION
**Pull request by jedireza**
@mcav and I worked on this.

### CODE REVIEW

File: src/context.rs [hunk 1]
```diff
@@ -35,6 +36,8 @@ pub trait ContextTrait {
               ws_port: Option<u16>) -> Shared<Self>;
     fn add_service(&mut self, service: Box<Service>);
     fn remove_service(&mut self, id: String);
+    fn add_socket(&mut self, socket_out: ws::Sender);
+    fn remove_socket(&mut self, id: ws::util::Token);
```
Comment (fabricedesre) [inline src/context.rs]: I would rather have remove_socket() also take a ws::Sender. No need to show the implementation detail used by add_socket()

File: src/controller.rs [hunk 2]
```diff
@@ -53,6 +57,11 @@ impl mio::Handler for Controller {
         println!("Receiving a notification! {}", data.description());
 
         let mut context = self.context.lock().unwrap();
+
+        for socket in context.sockets.values() {
+            let _ = socket.send(data.description());
```
Comment (fabricedesre) [inline src/controller.rs]: can we at least log when sending fail? And think about the right thing to do (eg. closing the socket?)

File: src/ws_server.rs [hunk 3]
```diff
@@ -0,0 +1,67 @@
+use std::thread;
```
Comment (fabricedesre) [inline src/ws_server.rs:5]: Please add the MPL2 header

File: src/context.rs (hunk 1 above)
Comment (jedireza) [inline src/context.rs]: updated

File: src/controller.rs (hunk 2 above)
Comment (jedireza) [inline src/controller.rs]: logging now

File: src/ws_server.rs (hunk 3 above)
Comment (jedireza) [inline src/ws_server.rs:5]: added

File: src/ws_server.rs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+use context::{ ContextTrait, SharedContext };
+use ws::{Handler, Sender, Result, Message, Handshake, CloseCode, Error};
```
Comment (fabricedesre) [inline src/ws_server.rs]: Nit: { Handler... , Error };

File: src/ws_server.rs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            let port;
+            {
+                let ctx = context.lock().unwrap();
+                hostname = ctx.hostname.clone();
+                port = ctx.ws_port;
+            }
+
+            let host_port = (&hostname as &str, port);
+
+            listen(host_port, |out| {
```
Comment (fabricedesre) [inline src/ws_server.rs]: nit: inline host_port here?

File: src/ws_server.rs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        println!("Hello new ws connection");
+
+        self.context.lock().unwrap().add_socket(self.out.clone());
+
+        Ok(())
+    }
+
+    fn on_message(&mut self, msg: Message) -> Result<()> {
+        // echo the message back to the client
+        self.out.send(msg)
```
Comment (fabricedesre) [inline src/ws_server.rs]: Not sure if this is a good idea to send back the message to the client. Let's just log for now.

File: src/ws_server.rs (hunk 4 above)
Comment (jedireza) [inline src/ws_server.rs]: fixed

File: src/ws_server.rs (hunk 5 above)
Comment (jedireza) [inline src/ws_server.rs]: fixed

File: src/ws_server.rs (hunk 6 above)
Comment (jedireza) [inline src/ws_server.rs]: logging now instead of echoing

File: src/context.rs [hunk 7]
```diff
@@ -12,15 +12,16 @@ use std::io::Error;
...
-
-    hostname: String,
-    http_port: u16,
-    ws_port: u16,
-    services: HashMap<String, Box<Service>>
+    pub hostname: String,
+    pub http_port: u16,
+    pub ws_port: u16,
+    pub services: HashMap<String, Box<Service>>,
+    pub sockets: HashMap<ws::util::Token, ws::Sender>,
```
Comment (fabricedesre) [inline src/context.rs]: it still looks like `services` and `sockets` can be private. I'd rather keep that.

File: src/context.rs [hunk 8]
```diff
@@ -72,6 +76,14 @@ impl ContextTrait for Context {
         self.services.remove(&id);
     }
 
+    fn add_socket(&mut self, socket: ws::Sender) {
```
Comment (fabricedesre) [inline src/context.rs]: Change to add{remove}_websocket(...). socket is a bit too generic.

File: src/context.rs (hunk 7 above)
Comment (jedireza) [inline src/context.rs]: I only made things public as needed.

File: src/context.rs [hunk 9]
```diff
@@ -84,6 +97,10 @@ impl ContextTrait for Context {
         serde_json::to_string(&self.services)
     }
 
+    fn websockets_iterator(&self) -> hash_map::Values<ws::util::Token, ws::Sender> {
```
Comment (fabricedesre) [inline src/context.rs]: sorry to be nitpicking again: the trend in Rust land seems to be to just use the short `iter` form (eg. https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.iter) so let's use websockets_iter(&self) here.

Comment (jedireza) [inline src/context.rs]: fixed

**Comment by fabricedesre** [comment]
r=me, please squash the commits before merging.

**Comment by jedireza** [comment]
:raised_hands:

**Pull request merged**